### ISSUE TITLE
[Experiment Report] 2025-09-06 ~ (김장원) 512+ pretrained models 변경 후 실험 계속

### ISSUE THREAD

**Issue by jkim1209** [AIBootcamp14/computervisioncompetition-cv2#30]
(서버이슈로 작업이 강제종료되어 코드 구조 소폭 수정됨에 따라 이슈 Close)

```cmd
The server was shut down due to excessive use of disk or memory. 
Due to the nature of the system, data recovery is not possible, so please create a new server and use it.
```

## 🧪 실험 개요

- 실험 이름: 512+ pretrained models 변경 후 실험 계속

- 실험 개요: 
  - 큰 틀은 이전실험과 비슷함: https://github.com/AIBootcamp14/computervisioncompetition-cv2/issues/26 
  - 잘못 label된 데이터 수정 후 작업: https://github.com/AIBootcamp14/computervisioncompetition-cv2/issues/29 
  - 모델별/데이터별 cv2 보간법 변경 후 실험: 개별 sample의 긴 변을 기준으로 square resize를 수행할 때 다음과 같은 interpolate 방식을 따름 (Auto)
    - default: `cv2.inter_linear`
    - image downsample하는 경우: `cv2.inter_area`
    - image upsample하는 경우: `cv2.inter_cubic`
  - LR 업데이트 스케쥴러 변경 후 실험 (CosineAnnealingWarmup)
  - 앙상블 기법 고도화
    - OOF 예측확률의 단순평균 뿐만 아니라 로그평균도 이용
    - OOF로 도출한 데이터 셋을 한번 더 CV 이용하여 최종 메타모델 생성하여 예측
    - 위의 메타모델과 별개로 OOF 확률의 기하평균을 이용한 예측값도 생성
  - 하이퍼파라미터를 변경해가며 계속 실험
  - 이후 변경사항:
    - 2025-09-07: 
      - Valid Set에 관련된 문제 수정: #31 
      - ~~early stopping을 Valid f1 score가 아닌 Valid Loss를 기준으로 변경~~
        - → Valid loss 와 Valid f1 score를 모두 고려하는 방향으로 변경 (둘 중 하나라도 좋아지면 patience 초기화)
      - 로깅 프린트 개선
      - Wandb 기록 추가
      - transform 기법 개선 (transform.py):
        - 전체적인 변형 확률 강화 및 개선
        - 소수 class 문서류 및 자주 잘못 분류하는 문서류에 대한 강화기법 추가 (`make_tf_doc_heavy`)
        - 문서류에 대한 overlay 기법 고도화 (`OverlayFromPool`)
        - 재현성을 위한 코드 보강
     - 모델 하이퍼파라미터 변경:
       - EMA decay : 0.999 → 0.9995 → 0.9999
       - LabelSmoothing : 0.01 → 0.05
        
- 날짜: 2025-09-06 ~

- 사용 데이터셋: aistage 경진대회 데이터셋
 
- 사용 모델 및 하이퍼파라미터:  

| Case | arch | epochs | effective batch_size | early stopping | patience | min_delta_loss| min_delta_f1 | loss function | lr_optimizer | lr_scheduler | lr (max_lr) | n_folds <br> (StratifiedKFold) | img_size | cv2 interpolate method | TTA |
| :---: | --- | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| 1 | MaxViT-Base (512) | 200 | 16 | True | 10 | 1e-3 | 0 | LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 1e-4 | 5 | Auto(512) | Auto | True |
| 2 | ViT-SigLIP (512) | 200 | 32 | True | 10 | 1e-3 | 0 | LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 7e-5 | 5 | Auto(512) | Auto | True |
| 3 | TinyViT-21M (512) | 200 | 32 | True | 10 | 1e-3 | 0 |LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 1e-4 | 5 | Auto(512) | Auto | True |
| 4 | EfficientNet-B6 (528) | 200 | 32 | True | 10 | 1e-3 | 0 |LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 7e-5 | 5 | Auto(528) | Auto | True |
| 5 | ViT-Base DINOv2 (518) | 200 | 32 | True | 10 | 1e-3 | 0 |LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 7e-5 | 5 | Auto(518) | Auto | True |
| 6 | BEiT-Large (512) | 200 | 16 | True | 10 | 1e-3 | 0 |LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 5e-5 | 5 | Auto(512) | Auto | True |
| 7 | EfficientNet-B7 (600) | 200 | 32 | True | 12 | 2e-3 | 0 |LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 7e-5 | 5 | Auto(600) | Auto | True |
| 8 | ConvNeXtV2-Huge (512) | 200 | 16 | True | 12 | 2e-3 | 0 |LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 5e-5 | 5 | Auto(512) | Auto | True |
| 9 | VOLO-D5 (512) | 200 | 16 | True | 12 | 2e-3 | 0 |LabelSmoothingCE | AdamW | CosineAnnealingWarmup | 5e-5 | 5 | Auto(512) | Auto | True |

> effective batch_size = batch_size × accum_steps

---

## 📈 결과

- 성능 지표 (Macro F1 Score):

| Case | arch | Macro F1 Score (Valid, Each Fold)| Macro F1 Score (Test) |
| :---: | --- | :---: | :---: |
| baseline | prev. ens. case (1 + 2 + 3) | - | 0.9568 |
| 1 | MaxViT-Base (512) |  |  |
| 2 | ViT-SigLIP (512) |  |  |
| 3 | TinyViT-21M (512) |  |  |
| 4 | EfficientNet-B6 (528) |  |  |
| 5 | ViT-Base DINOv2 (518) |  |  |
| 6 | BEiT-Large (512) |  |  |
| 7 | EfficientNet-B7 (600) |  |  |
| 8 | ConvNeXtV2-Huge (512) |  |  |
| 9 | VOLO-D5 (512) |  |  |
| ens2 | case (1 + 2) | - | cv-based:     <br> geomean:      |
| ens3 | case (1 + 2 + 3)  | - | cv-based:     <br> geomean:      |
| ens3 | case (1 + 2 + 9)  | - | cv-based:     <br> geomean:      |
| ens4 | case (1 + 2 + 3 + 4)  | - | cv-based:     <br> geomean:      |
| ens4 | case (1 + 2 + 4 + 9)  | - | cv-based:     <br> geomean:      |
| ens5 | case (1 + 2+ 4 + 5 + 9) | - | cv-based:     <br> geomean:      |
| ens6 | case (1 + 2+ 4 + 5 + 6 + 9) | - | cv-based:     <br> geomean:      |
| ens7 | case (1 + 2+ 4 + 5 + 6 + 8 + 9) | - | cv-based:     <br> geomean:      |
| ens8 | case (1 + 2+ 4 + 5 + 6 + 7 + 8 + 9) | - | cv-based:     <br> geomean:      |
| ens9 | case (all) | - | cv-based:     <br> geomean:      |

> baseline은 이전 실험 최고기록: https://github.com/AIBootcamp14/computervisioncompetition-cv2/issues/26
> 앙상블은 모두 OOF(Out-Of-Fold) 예측확률을 이용하여 모델을 결합함. 구체적으로 다음의 두 가지 방법으로 진행:
> - 모델별 OOF 예측확률로부터 도출된 Train data에 대한 예측값을 CV를 이용하여 가중치를 선택 후 앙상블 → 메타모델 생성 후 Test 데이터에 대한 예측값 도출 : cv-based
> - 모델별 OOF 예측확률을 가중평균하여 앙상블 → 메타모델 생성 후 Test 데이터에 대한 예측값 도출 : geomean

---

## 📚 참고자료:
  - cv2 보간법 선택기준: https://gist.github.com/georgeblck/e3e0274d725c858ba98b1c36c14e2835
  - CosineAnnealingWarmup: 
    - https://stydy-sturdy.tistory.com/56
    - source code: https://github.com/katsura-jp/pytorch-cosine-annealing-with-warmup/blob/master/cosine_annealing_warmup/scheduler.py
